db/db_merge_operator_test.cc

```diff
@@ -949,6 +951,98 @@ TEST_P(PerConfigMergeOperatorPinningTest, Randomized) {
   VerifyDBFromMap(true_data);
 }
 
+TEST_F(DBMergeOperatorTest, MaxSuccessiveMergesBaseValues) {
+  Options options = CurrentOptions();
+  options.create_if_missing = true;
+  options.merge_operator = MergeOperators::CreatePutOperator();
+  options.max_successive_merges = 1;
+  options.env = env_;
+  Reopen(options);
+
+  constexpr char foo[] = "foo";
+  constexpr char bar[] = "bar";
+  constexpr char baz[] = "baz";
+  constexpr char qux[] = "qux";
+  constexpr char corge[] = "corge";
+
+  // No base value
+  {
+    constexpr char key[] = "key1";
+
+    ASSERT_OK(db_->Merge(WriteOptions(), db_->DefaultColumnFamily(), key, foo));
+    ASSERT_OK(db_->Merge(WriteOptions(), db_->DefaultColumnFamily(), key, bar));
+
+    PinnableSlice result;
+    ASSERT_OK(
+        db_->Get(ReadOptions(), db_->DefaultColumnFamily(), key, &result));
+    ASSERT_EQ(result, bar);
+
+    // We expect the second Merge to be converted to a Put because of
```

Comment:
nit: there's no conversion to happen in no base value case, right?

Reply:
Conversion can actually happen even if there's no base value. In this case, the application's merge operator gets called with "no existing value" plus the merge operands, and as usual can output a new value, a new entity, or an existing operand as a result. The test uses `PutOperatorV2`, which returns the last operand as a plain value, resulting in the second `Merge` getting turned into a `Put` (similarly to what happens in the plain base value case).

Reply:
Yup. So IIUC, in this no base value case, the first Merge was actually done as "Put" and the second Merge worked as a "Merge", right? (My question was regarding the comment in line 980 and 981)

Reply:
It's the second `Merge` that's converted to a `Put`, since that's when we hit `max_successive_merges`. Note that internally, key-values with the same user key are sorted by sequence number in *descending* order (latest first), so that's the order `GetAllKeyVersions` returns them.

Reply:
Ah! Thank you! Should have checked inside `GetAllKeyVersions()` before commenting. With `descending order (latest first)`, now it makes perfect sense including how the second merge becomes put even with the no base value case.